### PR TITLE
change: Update alertmanager version to v0.34.0 and change GoDuration to NonZeroGoDuration type

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request._

Update alertmanager version to v0.34.0

ref: 
https://github.com/prometheus/alertmanager/blob/main/app/options.go#L170-L182
https://github.com/prometheus/alertmanager/blob/main/app/options.go#L210-L229

Alertmanager v0.34.0 now rejects retention: 0 at startup (Options.Retention must be positive), causing the pod to never become ready.

If you're contributing for the first-time, check our [contribution guidelines](../CONTRIBUTING.md).

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [x] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [ ] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [x] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
None
```

### CODE REVIEW

**Comment by dongjiang1989** [comment]
cc @slashpai  PTAL, thanks

File: scripts/go.mod [hunk 1]
```diff
@@ -9,7 +9,7 @@ require (
 	github.com/golangci/golangci-lint/v2 v2.13.1
 	github.com/google/go-jsonnet v0.22.0
 	github.com/jsonnet-bundler/jsonnet-bundler v0.6.0
-	github.com/prometheus/prometheus v0.314.0
```
Comment (simonpasquier) [inline scripts/go.mod:12]: why the change?

File: pkg/alertmanager/statefulset_test.go [hunk 2]
```diff
@@ -983,11 +983,12 @@ func TestSHAAndTagAndVersion(t *testing.T) {
...
 func TestRetention(t *testing.T) {
 	tests := []struct {
-		specRetention     monitoringv1.GoDuration
-		expectedRetention monitoringv1.GoDuration
+		specRetention     monitoringv1.NonZeroGoDuration
+		expectedRetention monitoringv1.NonZeroGoDuration
 	}{
 		{"", "120h"},
 		{"1d", "1d"},
+		{"0", "0"},
```
Comment (simonpasquier) [inline pkg/alertmanager/statefulset_test.go]: no need to test the zero value if it's rejected by the API

**Review by simonpasquier — COMMENTED** [review]
it's technically an API change while zero values didn't lead to a crash of Alertmanager (at least for `--data.retention`) on versions < v0.34.0. I'd rather go the safest route of not changing the API but failing the reconciliation when zero values are defined and version >= v0.34.0. We can revisit later the API change.

File: pkg/alertmanager/statefulset_test.go (hunk 2 above)
Comment (dongjiang1989) [inline pkg/alertmanager/statefulset_test.go]:
```suggestion
```

File: scripts/go.mod (hunk 1 above)
Comment (dongjiang1989) [reply scripts/go.mod:12]: I need rebase

**Comment by dongjiang1989** [comment]
@simonpasquier  git rebase done.

**Comment by slashpai** [comment]
> it's technically an API change while zero values didn't lead to a crash of Alertmanager (at least for `--data.retention`) on versions < v0.34.0. I'd rather go the safest route of not changing the API but failing the reconciliation when zero values are defined and version >= v0.34.0. We can revisit later the API change.


Since working on v0.94 release today and to bump alertmanager to v0.34.0 I will address this in separate PR and we can keep this PR for revisiting API change
